source/TypeTopology/AbsolutenessOfCompactness.lagda

```diff
@@ -0,0 +1,403 @@
+Andrew Swan, February 7th 2024
+
+This is a generalisation of some of the results by Martín Escardó in
+TypeTopology.PropTychonoff, based on the observation that for
+propositions P, the functor sending A to P → A is a
+modality. Modalities of this form are an important special case and
+they have a name; they are *open modalities* (Example 1.7 in
+[1]). However, we will show a version of the theorem is not only true
+for open modalities, but for all modalities.
+
+For another example, let ∇ be the modality of double negation sheaves
+(Example 3.41 of [1]). The internal logic in this reflective universe
+is boolean. It follows that ∇ (is-compact∙ A) holds for all types A,
+and so we can deduce that ∇ A is always compact.
+
+TODO: add double negation sheaves to the library
+
+We can also see as a special case that truncation preserves
+compactness, although it seems unlikely there are any good examples of
+compact higher types where it isn't already clear that the
+0-truncation is compact.
+
+When formulated in terms of modalities, the result is best thought of
+as an "absoluteness result." When working in models of some theory, a
+logical formula might make sense stated both internally in the model
+and for the same object viewed externally from outside the
+model. Absoluteness says that these two statements are equivalent.
+
+In this case we are thinking of reflective subuniverses as models of
+type theory sitting inside some larger "external" model of type
+theory. We will show that compactness is an upwards absolute
+notion. That is, if a type inside the reflective subuniverse is
+compact with respect to the internal logic of the subuniverse then it
+is compact viewed outside the subuniverse as just a type. The converse
+does not quite hold, so potentially there could be compact types where
+the internal statement of compactness is not true, and we don't get
+full absoluteness.
```

Comment:
Interesting. Maybe you can add a TODO to think about a potential counter-example?

But I am a bit confused here. The type N of natural numbers is not compact in many models, including topological models, but the type (LPO -> N) is compact.

Reply:
I don't have a clean Brouwerian counterexample that I can easily formalise in type theory and add, but I think this can happen in realizability models. In the effective topos the real numbers are precisely the computable reals and they are vacuously compact, because every function R -> 2 is constant. There's then a modality that "forces the halting set to be decidable." This adds non constant functions from the computable reals to 2 that prevents them from being compact in the subtopos, but I think the computable reals of the subtopos would still be compact in the effective topos, if that makes sense.